packages/pdf-viewer/VerticalPages.tsx

```diff
@@ -1,42 +1,72 @@
-import React, { useLayoutEffect, useRef, useState } from 'react';
+import { useLayoutEffect, useEffect, useRef, useState, MutableRefObject } from 'react';
+import * as React from 'react';
 import { PdfData, ScaledSize } from './pdfSource';
 import Page from './Page';
+import useAsyncEffect, { AsyncEffectEvent } from '@joplin/lib/hooks/useAsyncEffect';
+import styled from 'styled-components';
 
-require('./pages.css');
 
+const PagesHolder = styled.div<{ pageGap: number }>`
+	display: flex;
+	justify-content: center;
+	align-items: center;
+	flex-flow: column;
+	width: 100%;
+	min-height: 100%;
+	row-gap: ${(props)=> props.pageGap || 2}px;
+`;
 
 export interface VerticalPagesProps {
 	pdf: PdfData;
 	isDarkTheme: boolean;
-	anchorPage: number;
-	container: React.MutableRefObject<HTMLElement>;
+	anchorPage?: number;
+	rememberScroll?: boolean;
+	pdfId?: string;
+	container: MutableRefObject<HTMLElement>;
+	pageGap?: number;
+	showPageNumbers?: boolean;
 }
 
 
 export default function VerticalPages(props: VerticalPagesProps) {
 	const [scaledSize, setScaledSize] = useState<ScaledSize>(null);
+	const currentScaleSize = useRef(scaledSize);
 	const innerContainerEl = useRef<HTMLDivElement>(null);
+
+	const updateSize = async (cancelled?: boolean) => {
+		if (props.pdf) {
+			const containerWidth = innerContainerEl.current.clientWidth - 10;
+			const scaledSize_ = await props.pdf.getScaledSize(null, containerWidth);
+			if (cancelled) return;
+			innerContainerEl.current.style.height = `${(scaledSize_.height + (props.pageGap || 2)) * props.pdf.pageCount}px`;
+			const oldScaleSize = currentScaleSize.current;
+			if (oldScaleSize && props.container.current) {
+				const oldScrollTop = props.container.current.scrollTop;
+				props.container.current.scrollTop = oldScrollTop * (scaledSize_.scale / oldScaleSize.scale);
+			}
+			currentScaleSize.current = scaledSize_;
+			setScaledSize(scaledSize_);
+		}
+	};
+
 	useLayoutEffect(() => {
 		let resizeTimer: number = null;
 		let cancelled = false;
-		const updateSize = async () => {
-			if (props.pdf) {
-				const innerWidth = innerContainerEl.current.clientWidth;
-				const scaledSize = await props.pdf.getScaledSize(null, innerWidth - 10);
-				if (cancelled) return;
-				setScaledSize(scaledSize);
-			}
+
+		const _updateSize = async () => {
+			return updateSize(cancelled);
 		};
+
 		const onResize = () => {
 			if (resizeTimer) {
 				clearTimeout(resizeTimer);
 				resizeTimer = null;
 			}
-			resizeTimer = window.setTimeout(updateSize, 200);
+			resizeTimer = window.setTimeout(_updateSize, 200);
```

Comment:
This whole logic is wrong but can't wrap my head around it. You call updateSize from a timer, so if the effect is cancelled updateSize won't be notified and will keep running

Reply:
You mean the case where `updateSize` has already been called and `cancelled` becomes true after that?
For this I think we can pass a func instead of `cancelled` to `updateSize` like: `isCancelled` returning the current state each time. 
Does this sound better?

Reply:
I don't know. The thing with state is that the more level of indirections you have, the harder it is to know what's correct or not. That's why state logic is usually wrapped in a custom hook, which makes it easy to see what's happening. With your timers and intermediate functions it's very difficult to reason about the code.